src/test/java/com/cdancy/jenkins/rest/features/QueueApiLiveTest.java

```diff
@@ -57,6 +59,87 @@ public void testGetQueue() {
         assertTrue(foundLastKickedJob);
     }
 
+    @Test
+    public void testGetPendingQueueItem() {
+        Integer job1 = api.jobsApi().build("QueueTest");
+        assertNotNull(job1);
+        Integer job2 = api.jobsApi().build("QueueTest");
+        assertNotNull(job2);
+
+        // job2 is queue after job1, so while job1 runs, job2 is pending in the queue
+        QueueItem queueItem = api().queueItem(job2);
+        assertFalse(queueItem.cancelled());
+        assertNotNull(queueItem.why());
+        assertTrue(queueItem.why() instanceof String);
+        assertNull(queueItem.executable());
+    }
+
+    @Test
+    public void testGetRunningQueueItem() throws InterruptedException {
+        Integer job1 = api.jobsApi().build("QueueTest");
+        assertNotNull(job1);
+        Integer job2 = api.jobsApi().build("QueueTest");
+        assertNotNull(job2);
+
+        // job1 runs first, so we get its queueItem
+        QueueItem queueItem = getRunningQueueItem(job1);
+
+        // If null, it means the queueItem has been cancelled, which would not be normal in this test
+        assertNotNull(queueItem);
+        assertFalse(queueItem.cancelled());
+
+        //  We exepect this build to run, consequently:
+        //  * the why field should now be null
+        //  * the executable field should NOT be null
+        //  * the build number should be set to an integer
+        //  * the url for the build should be set to a string
+        assertNull(queueItem.why());
+        assertNotNull(queueItem.executable());
+        assertTrue(queueItem.executable().number() instanceof Integer);
+        assertTrue(queueItem.executable().url() instanceof String);
+    }
+
+    @Test
+    public void testGetCancelledQueueItem() throws InterruptedException {
+        Integer job1 = api.jobsApi().build("QueueTest");
+        assertNotNull(job1);
+        Integer job2 = api.jobsApi().build("QueueTest");
+        assertNotNull(job2);
+
+        RequestStatus success = api().cancel(job2);
+        assertTrue(success.value());
```

Comment:
@cdancy. This assert never succeeds. The log says there is a redirection (302) and I am not sure how to deal with that...
```
    21:59:15.061 [Test worker] DEBUG org.jclouds.rest.internal.InvokeHttpMethod - >> invoking queue:cancel
    21:59:15.061 [Test worker] DEBUG org.jclouds.http.internal.JavaUrlHttpCommandExecutorService - Sending request 830640781: POST http://localhost:8082/queue/cancelItem HTTP/1.1
    21:59:15.062 [Test worker] DEBUG jclouds.wire - >> "id=257"
    21:59:15.062 [Test worker] DEBUG jclouds.headers - >> POST http://localhost:8082/queue/cancelItem HTTP/1.1
    21:59:15.062 [Test worker] DEBUG jclouds.headers - >> Accept: application/json
    21:59:15.062 [Test worker] DEBUG jclouds.headers - >> Authorization: Basic bWFydGluOm1hcnRpbg==
    21:59:15.062 [Test worker] DEBUG jclouds.headers - >> Jenkins-Crumb: 53842b63c40e494a74d261a4773339ef
    21:59:15.062 [Test worker] DEBUG jclouds.headers - >> Content-Type: application/x-www-form-urlencoded
    21:59:15.062 [Test worker] DEBUG jclouds.headers - >> Content-Length: 6
    21:59:15.157 [Test worker] DEBUG org.jclouds.http.internal.JavaUrlHttpCommandExecutorService - Receiving response 830640781: HTTP/1.1 302 Found
    21:59:15.157 [Test worker] DEBUG jclouds.headers - << HTTP/1.1 302 Found
    21:59:15.157 [Test worker] DEBUG jclouds.headers - << Server: Jetty(9.4.z-SNAPSHOT)
    21:59:15.157 [Test worker] DEBUG jclouds.headers - << X-Content-Type-Options: nosniff
    21:59:15.157 [Test worker] DEBUG jclouds.headers - << Date: Thu, 19 Apr 2018 01:59:15 GMT
    21:59:15.157 [Test worker] DEBUG jclouds.headers - << Location: http://localhost:8082/queue/
    21:59:15.157 [Test worker] DEBUG jclouds.headers - << Content-Type: application/unknown
    21:59:15.157 [Test worker] DEBUG jclouds.headers - << Content-Length: 0
    21:59:15.158 [Test worker] DEBUG org.jclouds.http.internal.JavaUrlHttpCommandExecutorService - Sending request -1390855584: POST http://localhost:8082/queue/ HTTP/1.1
    21:59:15.158 [Test worker] DEBUG jclouds.wire - >> "id=257"
    21:59:15.158 [Test worker] DEBUG jclouds.headers - >> POST http://localhost:8082/queue/ HTTP/1.1
    21:59:15.158 [Test worker] DEBUG jclouds.headers - >> Accept: application/json
    21:59:15.158 [Test worker] DEBUG jclouds.headers - >> Authorization: Basic bWFydGluOm1hcnRpbg==
    21:59:15.158 [Test worker] DEBUG jclouds.headers - >> Jenkins-Crumb: 53842b63c40e494a74d261a4773339ef
    21:59:15.159 [Test worker] DEBUG jclouds.headers - >> Content-Type: application/x-www-form-urlencoded
    21:59:15.159 [Test worker] DEBUG jclouds.headers - >> Content-Length: 6
    21:59:15.252 [Test worker] DEBUG org.jclouds.http.internal.JavaUrlHttpCommandExecutorService - Receiving response -1390855584: HTTP/1.1 404 Not Found
    21:59:15.253 [Test worker] DEBUG jclouds.headers - << HTTP/1.1 404 Not Found
    21:59:15.253 [Test worker] DEBUG jclouds.headers - << Server: Jetty(9.4.z-SNAPSHOT)
    21:59:15.253 [Test worker] DEBUG jclouds.headers - << X-Content-Type-Options: nosniff
    21:59:15.253 [Test worker] DEBUG jclouds.headers - << Date: Thu, 19 Apr 2018 01:59:15 GMT
```

Reply:
Found that the redirect to a non-existent page is actually issue [JENKINS-21311](https://issues.jenkins-ci.org/browse/JENKINS-21311)